doc/manual/man/man1/extract_seq.1
.TH "extract_seq" 1 "" "" "Staden Package"
.SH "NAME"
.PP
extract_seq \- extracts sequence from a trace or experiment file.

.SH "SYNOPSIS"
.PP

\fBextract_seq\fP [\fB-r\fP]
[\fB-\fP(\fBabi\fP|\fBalf\fP|\fBscf\fP|\fBztr\fP|\fBexp\fP|\fBpln\fP)]
[\fB-good_only\fP] [\fB-clip_cosmid\fP] [\fB-fasta_out\fP]
[\fB-output\fP \fIoutput_name\fP] [\fIinput_name\fP] \fB...\fP

.SH "DESCRIPTION"
.PP

\fBextract_seq\fP extracts the sequence information from binary trace
files, Experiment files, or from the old Staden format plain files. The input
can be read either from files or from standard input, and the output can be
written to either a file or standard output. Multiple input files can be
specified. The output contains the sequences split onto lines of at most 60
characters each.

.SH "OPTIONS"
.PP

.TP
\fB-r\fP
Directs reading of experiment file to attempt extraction of sequence from
the referenced (\fBLN\fP and \fBLT\fP line types) trace file. Without
this option, or when the trace file cannot be found, the sequence
output is that listed in the Experiment File. This option has no effect
for other input format types.

.TP
\fB-abi\fP, \fB-alf\fP, \fB-scf\fP, \fB-ztr\fP, \fB-exp\fP, \fB-pln\fP
Specify an input file format. This is not usually required as
\fBextract_seq\fP will automatically determine the correct input file
type. This option is supplied incase the automatic determination is
incorrect (which is possible, but has never been observed).

.TP
\fB-good_only\fP
When reading an experiment file or SCF file containing clip marks, output
only the \fIgood\fP sequence which is contained within the boundaries marked
by the \fBQL\fP, \fBQR\fP, \fBSL\fP, \fBSR\fP, \fBCL\fP, \fBCR\fP
and \fBCS\fP line types.

.TP
\fB-clip_cosmid\fP
When the \fB-good_only\fP argument is specified this controls whether the
cosmid sequence should be considered good data. Without this argument
cosmid sequence is considered good.

.TP
\fB-fasta_out\fP
Specifies that the output should be in fasta format

.TP
\fB-output\fP \fIfile\fP
The sequence will be written to \fIfile\fP instead of standard
output.
.TE
.SH "SEE ALSO"
.PP

\fBExperimentFile\fR(4)
\fBscf\fR(4)
_fxref(Man-extract_fastq extract_fastq(1), extract_fastq.1)
\fBRead\fP(4)


NEWS
-----------------------------------------------------------------------------
12th April 2006
===============

Gap4/Trev
---------

* Solexa data should now be viewable using a stick-style plot.
  The 4 log-odds scaled confidence values are also now visible.

* Sequence Search can now search for matches listed in a file.

* Bug fixes to the Shuffle Pads algorithm.


-----------------------------------------------------------------------------
12th December 2005
==================

Gap4
----

* Find Internal Joins now ranks hits in the contig comparator by
  alignment score rather than percentage identity. This change does
  not alter the hits found; just their ordering.

* The Join Editor now as align left and align right buttons ("<" and
  ">" either side of "align"). These anchor one end of the alignment
  at the editor cursor position and allow the other end to move to
  accommodate padding. The old align button works as previously.

* Added an option to force the maximum number of groups formed in the
  SNP Candidates window, even if this means arbitrarily picking groups
  to merge. Also added protection against non-ACGT characters in sequences.

* Added a Save button to the "list" viewer/editor.

* The Contig Navigator window now has page-up and page-down
  keybindings and a new checkbutton to automatically display a
  selection of traces.

Misc
----

* Rewrote the startup scripts (gap4, spin, pregap4, trev, ...) to
  source staden.profile (only when appropriate) instead of duplicating
  much of the functionality there.

-----------------------------------------------------------------------------
15th November 2005
==================


Gap4
----

* Improved the SNP Candidates plot; it now generates consensus
  sequence per template when trying to identify candidate snps, rather
  than simply throwing out locations where conflicts exist within a
  template.

* Shuffle pads now has a configurable "band size" parameter. Also
  fixed a tag corruption bug in it.

* The contig editor shuffle pads command is back in a minimal form. It
  turns out that people rather liked its ability to remove entire
  columns of pads, so that particular feature of it is available once
  more.


io_lib
------

* Added append_sff program for merging 454's SFF archives.

* Various bug fixes with the mFILE support.

-----------------------------------------------------------------------------
20th October 2005
=================

Internal Sanger test version rel-1-6-0 ("gap4 -test")
(Ok so it's slightly newer than the official 1.6.0..)

Gap4
----

* SNP Candidates now has a 'correlation offset' parameter. This
  modifies the average correlation score required to give a positive
  impact on the edge scores.

io_lib
------

* Complete rewrite of the SFF format handling and replaced TVF with
  SFF archvies. New programs to manipulate SFF archives too: hash_sff
  and append_sff. Sorted and hash indices are supported in SFF archives.

* Added SFF= format for the RAWDATA search path.

* Allow archive/entry for use with SFF, TAR and HASH.
  Eg "extract_seq foo.tar/xyzzy.ztr" will work without setting
  RAWDATA.

-----------------------------------------------------------------------------
30th September 2005
===================

Version 1.6.0 tagged

-----------------------------------------------------------------------------

27th September 2005
===================

Gap4
----

* Renamed the discrepancy plot to be second highest confidence plot.

* Fixed some (apparently rare) find internal joins crashes.

* Added a minimum quality slider to the Report Mutations "by
  differences" mode. Also fixed crashes and missing pictures in the
  HTML report.

io_lib
------

* Changed the HashFile format and updated it to be official version
  1.00.

  The key difference is that there's now a footer pointing back into
  the hashfile header allowing for the index to simply be appended to
  a (e.g. tar) archive.

* Improved handling of 454 data in ZTR; it no longer attempts to
  compress raw pyrosequencing data as this was not working well.

* Fixed Bug ID 1289095 (file descriptor leak).

-----------------------------------------------------------------------------

24th August 2005
================

Gap4
----

* Rewrote the Shuffle Pads algorithm to be more like ReAligner. It can
  introduce new pads and remove old ones as necessary in order to
  optimise the alignments. It has also moved from inside the contig
  editor to part of the main menu so that it can be scripted easier.

* The join editor invoked from Find Read Pairs plot now brings up the
  editors at the start/end of the sequences forming the read-pair
  instead of the start/end of the two contigs.

* Search by sequence in the editor now finds consensus matches
  too. Also fixed some bugs where cutoff data matches were not always
  found and in some cases (reverse search) non-cutoff matches could be
  missed.

* The SNP Candidates plot now allows for the consensus to be saved
  directly from this window. The consensus and fake sequences added
  when a real split is requested now includes the "unsplit" readings
  from regions which apparently contained no SNPs (and therefore
  belong to all split contigs).

* Report Mutations bug fix. Producing HTML output now works (albeit
  with less content) when asking for SNPs to be derived from
  "differences" instead of tags.

Trev
----

* Pyrosequencing data (eg 454 flowgrams) are now supported. Trev
  auto-senses the type and views accordingly. Added format for 454's
  SFF standard too.

Io_lib
------

* New release - 1.9.0. No longer uses libtool in configuration.

* Added "HASH=hashfile" component of RAWDATA search path along with
  hash_tar and hash_extract tools. These operate like index_tar but
  provide true fast random access into tar files (or other formats if
  needed).

* Major update of I/O mechanisms to use in-memory equivalents of the
  files being read from or written to. This has lead to some

src/io_lib/progs/extract_seq.c
/*
 * Copyright (c) Medical Research Council 1994-1999. All rights reserved.
 *
 * Permission to use, copy, modify and distribute this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * this copyright and notice appears in all copies.
 *
 * This file was written by James Bonfield, Simon Dear, Rodger Staden,
 * as part of the Staden Package at the MRC Laboratory of Molecular
 * Biology, Hills Road, Cambridge, CB2 2QH, United Kingdom.
 *
 * MRC disclaims all warranties with regard to this software.
 */

#include <stdio.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include "Read.h"
#include "traceType.h"
#include "expFileIO.h"
#include "open_trace_file.h"

/* #include "stdio_hack.h" */

#define LINE_LENGTH 60

static int do_trans(mFILE *infp, char *in_file, FILE *outfp, int format,
		    int good_only, int clip_cosmid, int fasta_out) {
    Read *r;
    char *tmp_base;

    read_sections(READ_BASES);
    if (NULL == (r = mfread_reading(infp, in_file, format))) {
	fprintf(stderr, "Failed to read file '%s'\n", in_file);
	return 1;
    }

    tmp_base = r->base;

#ifdef IOLIB_EXP
    if (good_only && r->orig_trace_format == TT_EXP) {
	int left=0, right=r->NBases + 1, val, lval, rval;
	Exp_info *e = (Exp_info *)r->orig_trace;

	if (0 == exp_get_int(e, EFLT_SL, &val))
	    if (val > left)
		left = val;
	if (0 == exp_get_int(e, EFLT_QL, &val))
	    if (val > left)
		left = val;

	if (0 == exp_get_int(e, EFLT_SR, &val))
	    if (val < right)
		right = val;
	if (0 == exp_get_int(e, EFLT_QR, &val))
	    if (val < right)
		right = val;

	/* This is horrid - see gap seqInfo.c file for explaination */
	if (clip_cosmid) {
	    int got_cosmid;

	    if (0 == exp_get_rng(e, EFLT_CS, &lval, &rval)) {
		got_cosmid = 1;
	    } else if (0 == exp_get_int(e, EFLT_CL, &lval) &&
		       0 == exp_get_int(e, EFLT_CR, &rval)) {
		got_cosmid = 1;
	    } else {
		got_cosmid = 0;
	    }
	
	    if (got_cosmid) {
		if      (lval <= left   && rval <= left)    ;
		else if (lval <= left+1 && rval <  right)   left  = rval;
		else if (lval <= left+1 && rval >= right)   right = left+1;
		else if (lval <  right  && rval <  right)   right = lval;
		else if (lval <  right  && rval >= right)   right = lval;
	    }
	}

	r->base += left;
	r->NBases = right - left - 1;
    } else
#endif /* IOLIB_EXP */
    if (good_only) {
        r->base += r->leftCutoff;
	r->NBases = r->rightCutoff - r->leftCutoff - 1;
    }

    if (fasta_out) {
	char *p = strrchr(in_file, '/');
	int i;

	/* Add header */
	if (NULL == p)
	    p = in_file;
	else
	    p++;
	fprintf(outfp, ">%s\n", p); 

	/* Replace - with N */
	for (i = 0; i < r->NBases; i++) {
	    if (r->base[i] == '-')
		r->base[i] = 'N';
	}
    }
    fwrite_reading(outfp, r, TT_PLN);

    r->base = tmp_base;
    read_deallocate(r);
    fflush(outfp);

    return 0;
}

static void usage(void) {
    fprintf(stderr, "Usage: extract_seq [-r] [-(abi|alf|scf|exp|pln)]\n"
	    "                   [-good_only] [-clip_cosmid] [-fasta_out]\n"
	    "                   [-output output_name] [input_name] ...\n");
    exit(1);
}

int main(int argc, char **argv) {
    int from_stdin = 1;
    mFILE *infp = mstdin();
    FILE *outfp = stdout;
    int format = TT_ANY;
    int redirect = 1;
    int good_only = 0;
    int clip_cosmid = 0;
    int fasta_out = 0;
    int ret = 0;
    char *fofn = NULL;

    for (argc--, argv++; argc > 0; argc--, argv++) {
	if (strcmp(*argv, "-r") == 0) {
	    redirect = 0;
	} else if (strcasecmp(*argv, "-abi") == 0) {
            format = TT_ABI;
        } else if (strcasecmp(*argv, "-alf") == 0) {
            format = TT_ALF;
        } else if (strcasecmp(*argv, "-scf") == 0) {
            format = TT_SCF;
        } else if (strcasecmp(*argv, "-exp") == 0) {
            format = TT_EXP;
        } else if (strcasecmp(*argv, "-pln") == 0) {
            format = TT_PLN;
        } else if (strcasecmp(*argv, "-ztr") == 0) {
            format = TT_ZTR;
        } else if (strcasecmp(*argv, "-ctf") == 0) {
            format = TT_CTF;
        } else if (strcasecmp(*argv, "-good_only") == 0) {
	    good_only = 1;
        } else if (strcasecmp(*argv, "-clip_cosmid") == 0) {
	    clip_cosmid = 1;
	} else if (strcasecmp(*argv, "-fasta_out") == 0) {
	    fasta_out = 1;
	} else if (strcmp(*argv, "-fofn") == 0) {
	    fofn = *++argv;
	    argc--;
	    from_stdin = 0;
        } else if (strcasecmp(*argv, "-output") == 0) {
	    if (NULL == (outfp = fopen(*++argv, "wb"))) {
		perror(*argv);
		return 1;
	    }
            argc--;
	} else if (**argv != '-') {
	    from_stdin = 0;
	    break;
        } else {
            usage();
        }
    }

    read_experiment_redirect(redirect);

    if (!from_stdin) {
	if (fofn) {
	    FILE *fofn_fp;
	    char line[8192];

	    if (strcmp(fofn, "stdin") == 0)
		fofn_fp = stdin;
	    else
		fofn_fp = fopen(fofn, "r");

	    if (fofn_fp) {
		while (fgets(line, 8192, fofn_fp) != NULL) {
		    char *cp;
		    if (cp = strchr(line, '\n'))
			*cp = 0;
		    if (format == TT_EXP) {
			infp = open_exp_mfile(line, NULL);
		    } else {
			infp = open_trace_mfile(line, NULL);
		    }
		    if (NULL == infp) {
			perror(line);
			ret = 1;
		    } else {
			ret |= do_trans(infp, line, outfp, format, good_only,
					clip_cosmid, fasta_out);
			mfclose(infp);
		    }
		}
		fclose(fofn_fp);
	    }
	}
	for (;argc > 0; argc--, argv++) {
	    if (format == TT_EXP) {
		infp = open_exp_mfile(*argv, NULL);
	    } else {
		infp = open_trace_mfile(*argv, NULL);
	    }
	    if (NULL == infp) {
		perror(*argv);
		ret = 1;
	    } else {
		ret |= do_trans(infp, *argv, outfp, format, good_only,
				clip_cosmid, fasta_out);
		mfclose(infp);
	    }
	}
    } else {
	ret = do_trans(infp, "<stdin>", outfp, format, good_only, clip_cosmid,
		       fasta_out);
    }

    return ret;
}

